src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Metadata/ModelAttributes.cs

```diff
@@ -131,6 +155,23 @@ public static ModelAttributes GetAttributesForType(Type type)
             return new ModelAttributes(attributes);
         }
 
+        public static ModelAttributes GetAttributesForParameter(ModelMetadataIdentity key)
+        {
+            var parameter = key.ActionDescriptor?.Parameters.Single(p => p.Name.Equals(key.Name, StringComparison.Ordinal));
+            IEnumerable<object> parameterAttributes;
+
+            if (parameter is ControllerParameterDescriptor controllerParameter)
```

Comment:
TBH I'm not clear on the distinction between `ParameterDescriptor` and `ControllerParameterDescriptor`. I put the new stuff into `ControllerParameterDescriptor` just to minimise the change footprint. Any opinions on this?

Reply:
I'd rather see this API deal with reflection types only if possible.

In general a `ParameterDescriptor` represents the **idea** of a parameter and a `ControllerParameterDescriptor` represents an actual method parameter on an actual action method. So anything you compute from reflection should go on the `ControllerParameterDescriptor`. 

Do we need to add the attributes as a property here? I would think that `ControllerParameterDescriptor` exposes the `ParameterInfo` already

Reply:
OK. We need to preserve the actual attribute instances, not the reflection-only descriptors given by `ParameterInfo` (because, for example, the Data Annotations validation provider is looking for those actual instances).

Since this is *not* reflection-only info, it should then go on to `ParameterDescriptor` rather than `ControllerParameterDescriptor`, should it? I'll put it there now but let me know if you think I'm misreading your distinction.

Reply:
Update: Yes we can get the actual attribute instances from `ControllerParameterDescriptor`'s `ParameterInfo`. So we don't need the new `Attributes` property.